### ISSUE TITLE
Support for complex types of state values

### ISSUE THREAD

**Issue by martinsura** [avkonst/hookstate#106]
Hello, 

First - thanks for this nice library. 

I have one problem with state and Date.

Simple component:
```
export function DateComponent() {
  const date = useState<Date>(new Date());
  return <div>{date.get().toString()}</div>;
}
```
Returns 
```
TypeError
_this.child(...).get is not a function
```
Problem is default value (new Date())

Is this bug or I miss something?

Btw after some investigation - 
```
 date.attach(Downgraded);
```
works fine.

Thanks

**Comment by starlabs007** [comment]
@martinsura You can't use `Date` as a state value (`Maps`, `Sets` and `Dates` can't be used).
See here: https://hookstate.js.org/docs/nested-state#limitations-for-state-values

**Comment by martinsura** [comment]
Oh, sorry - missed this doc. 

Sorry for unnecessary issue.

Thanks for link

**Comment by starlabs007** [comment]
No problem; it's pretty easy to miss as it's a small section at the end of a very long page.

@avkonst Perhaps move the limitations section to its own category in the docs?

**Comment by avkonst** [comment]
Ok. I will check how to improve this doc. I have also experimented with adding support for Date and other complex JS types. It is possible to make it work for the @hookstate/core. But it is a problem for some plugins, as complex JS types are not JSON serializable - which is a weakness of the JS language itself. Not sure if adding this support is valuable. Until now I managed to work around this language limitation using plain strings...

**Comment by avkonst** [comment]
See also: https://github.com/avkonst/hookstate/issues/124

**Comment by channeladam** [comment]
This is a really great library that I want to use, but lack of support for Date is the only reason I will and cannot use it. :(

**Comment by avkonst** [comment]
The future major version will get automatic Downgrading for standard types. In the meantime you still can use the library with the Date object, no problems. It is just an inconvenience that you have to write `.attach(Downgraded).get()` instead of just `.get()` when you access Date type property. Otherwise, it is no issues to use Date.

**Comment by avkonst** [comment]
implemented in Hookstate 4.0.0-rc7
